src/boot/efi/util.c

```diff
@@ -748,20 +748,149 @@ EFI_STATUS device_path_to_str(const EFI_DEVICE_PATH *dp, char16_t **ret) {
+        uint8_t length;
+        uint8_t handle[2];
+} _packed_ SmbiosHeader;
+
+typedef struct {
+        SmbiosHeader header;
+        uint8_t vendor;
+        uint8_t bios_version;
+        uint16_t bios_segment;
+        uint8_t bios_release_date;
+        uint8_t bios_size;
+        uint64_t bios_characteristics;
+        uint8_t bios_characteristics_ext[2];
+} _packed_ SmbiosTableType0;
+
+static bool smbios_in_hypervisor(void) {
+        char *p;
+        uint64_t size = 0;
+
+        Smbios3EntryPoint *entry3 = find_configuration_table(&(EFI_GUID) SMBIOS3_TABLE_GUID);
+        if (entry3 && memcmp(entry3->anchor_string, "_SM3_", 5) == 0 &&
+            entry3->entry_point_length <= sizeof(*entry3)) {
+                p = PHYSICAL_ADDRESS_TO_POINTER(entry3->table_address);
+                size = entry3->table_maximum_size;
+        }
+
+        if (size == 0) {
+                SmbiosEntryPoint *entry = find_configuration_table(&(EFI_GUID) SMBIOS_TABLE_GUID);
+                if (entry && memcmp(entry->anchor_string, "_SM_", 4) == 0 &&
+                    entry->entry_point_length <= sizeof(*entry)) {
+                        p = PHYSICAL_ADDRESS_TO_POINTER(entry->table_address);
+                        size = entry->table_length;
+                }
+        }
+
+        /* Look up BIOS Information (Type 0). */
+        for (;;) {
+                if (size < sizeof(SmbiosHeader))
+                        return false;
+
+                SmbiosHeader *header = (SmbiosHeader *) p;
+
+                /* End of table. */
+                if (header->type == 127)
+                        return false;
+
+                if (size < header->length)
+                        return false;
+
+                if (header->type == 0)
+                        break; /* Yay! */
+
+                /* Skip over formatted area. */
+                size -= header->length;
+                p += header->length;
+
+                /* Skip over string table. */
+                for (;;) {
+                        while (size > 0 && *p != '\0') {
+                                p++;
+                                size--;
+                        }
+                        if (size == 0)
+                                return false;
+                        p++;
+                        size--;
+
+                        /* Double NUL terminates string table. */
+                        if (*p == '\0') {
+                                if (size == 0)
+                                        return false;
+                                p++;
+                                break;
+                        }
+                }
+        }
+
+        SmbiosTableType0 *type0 = (SmbiosTableType0 *) p;
+        if (type0->header.length < sizeof(SmbiosTableType0))
+                return false;
+
+        /* Bit 4 of 2nd BIOS characteristics extension bytes indicates virtualization. */
+        return FLAGS_SET(type0->bios_characteristics_ext[1], 1 << 4);
+}
```

Comment:
i would probably have split this into three functions, i.e. one that finds the table address/size, and another one which then finds the right type in it. and the one that glue this together.